pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java

```diff
@@ -176,76 +196,406 @@ public void shutdown() {
+    if (lastKnownInstanceConfigs == null || lastKnownInstanceConfigs.isEmpty()) {
+      LOGGER.info("Routing table for table {} requires rebuild due to empty/null previous instance configs", tableName);
+      return true;
+    }
+
+    // Gather relevant incoming instance configs
+    Map<String, InstanceConfig> currentRelevantInstanceConfigs = new HashMap<>();
+    for (InstanceConfig incomingInstanceConfig : instanceConfigs) {
+      String instanceName = incomingInstanceConfig.getInstanceName();
+
+      if (lastKnownInstanceConfigs.containsKey(instanceName)) {
+        currentRelevantInstanceConfigs.put(instanceName, incomingInstanceConfig);
       }
+    }
 
+    // Did some instances lose their configuration?
+    if (lastKnownInstanceConfigs.size() != currentRelevantInstanceConfigs.size()) {
       LOGGER.info(
-          "Updating routing table for table {} due to ZK change (current version {}, last known version {})",
-          tableName, externalViewRecordVersion, lastKnownZkVersion);
+          "Routing table for table {} requires rebuild due to having a different number of instance configs (known instance config count {}, current instance config count {})",
+          tableName, lastKnownInstanceConfigs.size(), currentRelevantInstanceConfigs.size());
+      return true;
+    }
+
+    // Did some instance change state?
+    for (String instanceName : lastKnownInstanceConfigs.keySet()) {
+      InstanceConfig previousInstanceConfig = lastKnownInstanceConfigs.get(instanceName);
+      InstanceConfig currentInstanceConfig = currentRelevantInstanceConfigs.get(instanceName);
+
+      // If it's the same znode, don't bother comparing the contents of the instance configs
+      if (previousInstanceConfig.getRecord().getVersion() == currentInstanceConfig.getRecord().getVersion()) {
+        continue;
+      }
+
+      // Check if the instance got enabled/disabled or started/stopped shutting down since the last update
+      boolean wasEnabled = previousInstanceConfig.getInstanceEnabled();
+      boolean isEnabled = currentInstanceConfig.getInstanceEnabled();
+
+      String wasShuttingDown =
+          previousInstanceConfig.getRecord().getSimpleField(CommonConstants.Helix.IS_SHUTDOWN_IN_PROGRESS);
+      String isShuttingDown =
+          currentInstanceConfig.getRecord().getSimpleField(CommonConstants.Helix.IS_SHUTDOWN_IN_PROGRESS);
+
+      boolean instancesChanged =
+          !EqualityUtils.isEqual(wasEnabled, isEnabled) || !EqualityUtils.isEqual(wasShuttingDown, isShuttingDown);
+
+      if (instancesChanged) {
+        LOGGER.info(
+            "Routing table for table {} requires rebuild due to at least one instance changing state (instance {} enabled: {} -> {}; shutting down {} -> {})",
+            tableName, instanceName, wasEnabled, isEnabled, wasShuttingDown, isShuttingDown);
+        return true;
+      } else {
+        // Update the instance config in our last known instance config, since it hasn't changed
```

Comment:
If the newly seen instance hasn't changed in a way that is considered different (eg. some field that we don't care about), then we keep it as the last seen version.